### DISCUSSION TITLE
Migrating to Nexus for Project Dependencies

### DISCUSSION THREAD

**Discussion by cristianvalentincojocaru** [eclipse-tycho/tycho#3371] [Q&A]
**TL;DR**
Is it possible to resolve Eclipse plugins and features dependencies using a Maven repository stored on Nexus? If yes, what is the recommended approach and the necessary steps to make this integration work effectively?

Hello everyone,

We're in the process of migrating to a Nexus storage solution for our project, and we've encountered challenges resolving project dependencies, especially for repositories with dependencies from Nexus.
Previously, we specified individual dependencies (maven repositories, p2 repositories, etc.) in the pom.xml for each repository, as seen in Example 1. This is no longer viable thus we are trying to streamline this process and achieve something similar to Eclipse's approach (Example in Eclipse JDT - https://github.com/eclipse-jdt/eclipse.jdt/blob/master/pom.xml#L48 ).

**Example 1:**
```
<repositories>
	<repository>
		<id>pack_1</id>
		<layout>p2</layout>
		<url>ftp://${ftp-user}:${ftp-password}/folder1/pack_1/</url>
	</repository>
	<!-- More repositories with similar structure -->
	<repository>
		<id>eclipse-updates-4.20</id>
		<layout>p2</layout>
		<url>https://download.eclipse.org/eclipse/updates/4.20/</url>
	</repository>
</repositories>
```

We managed to make it work with the following setup, but it involves a considerable amount of manual work to replace dependencies from all repositories:
```
<repositories>
	<repository>
		<releases>
			<enabled>true</enabled>
		</releases>
		<snapshots>
			<enabled>true</enabled>
		</snapshots>
		<id>snapshots</id>
		<url>https://company.nexus.com/repository/snapshots/</url>
	</repository>
	<repository>
		<id>pack_1</id>
		<url>mvn:pack_1.groupId:pack_1.artifactId:1.0.0-SNAPSHOT:zip</url>
		<layout>p2</layout>
	</repository>
	<!-- More repositories with similar structure -->
</repositories>
```

**Our goal:**
```
<repositories>
	<repository>
		<releases>
			<enabled>true</enabled>
		</releases>
		<snapshots>
			<enabled>true</enabled>
		</snapshots>
		<id>snapshots</id>
		<url>https://company.nexus.com/repository/snapshots/</url>
	</repository>
	<!-- Potentially other external repositories if needed -->
</repositories>
```

We would gladly appreciate any help and insights on how Eclipse manages to have such a streamlined process and solve all the dependencies using a single repository.

Additional details:
- Tycho version: 2.2.0
- JDK: 11.0.17

Thank you!

**Comment by laeubi** [comment]
> Tycho version: 2.2.0

The very first thing is you should really upgrade to the latest release (that is Tycho 4.0.4)

Beside that its a bit unclear what exact problem your are facing and why you have chosen the given structures, so you need at least to share an example of a project to get better answer.

If you require dedicated support you can find Tycho support options here:
- https://github.com/eclipse-tycho/tycho#getting-support

**Reply by cristianvalentincojocaru** [reply]
The main scenario would be this :
 
We have maven repos with different dependencies 
- other maven repos maintained by us, uploaded on Nexus.We specify them using Nexus rest API.
- other third party maven repos, with different binaries, which are currently not stored on Nexus, but rather on a FTP
 
What I was trying to do is to move the third party maven repos also on Nexus, move all repos in a single folder (let's say snapshots) and then reference in the repository area only the snapshots folder from Nexus.
 
We tried that and unfortunately maven doesn't resolve the features and plugins, if we specify only the root folder from nexus (snapshot in our case).

**Reply by laeubi** [reply]
I think your are mixing things up a bit, please note that "repository" is ambiguous term in tycho/maven.

- In maven a repository of layout `default` is https://maven.apache.org/repositories/layout.html intensified by GAVs (e.g. in a pom.xml)
- Then Tycho support also repositories of layout `p2` what is: https://wiki.eclipse.org/Equinox/p2/p2_index identified by ID+Version and can be quiried for packages as well.
- Tycho also supports some other layouts but I'll leave them out here.

The point is that none of those are a "folder" you can put on a Nexus or ftp server... and you can't "reference" then bu just a plain URL how should maven/tycho know the structure.

As said Tycho can use many types of repository layouts but then you need to configure them appropriately and maybe adjust your build (e.g. maven layout will require either maven-target definition or pom.xml configuration) and then it could work to host them all in a Nexus (repository Manager) under one URL.... but throing a random folder somewhere most probably will not work as intended.

**Reply by cristianvalentincojocaru** [reply]
Sorry for not offering a complete picture, We are not trying to upload a simple folder.
We managed to deploy some packages that do not require any internal dependencies. These packages are made up of an eclipse-plugin, eclipse feature and an eclipse repository, each including the other. These have been uploaded on Nexus, in a maven2 repository format, using the "deploy" maven lifecycle. 
The main question is how can we use the artifacts uploaded on nexus to solve the dependency of these packages in other projects?

**Reply by laeubi** [reply]
You can add them to the pom.xml as maven dependency or use m2e target locations and a target file.
